### DISCUSSION TITLE
Enumeration of users via auth mechanisms/prompts

### DISCUSSION THREAD

**Discussion by yaleman** [kanidm/kanidm#610] [General]
The auth flows allow for trivial account enumeration through the first step in auth, and should return "invalid auth" regardless of user/pass entered. This leads to multiple vectors for organisational attack, through either direct auth attack or social engineering.

This would require a change to the auth flow, currently is `user -> MFA -> password`, which enables this enumeration attack - IMO, it probably should _really_ be user + password, then MFA.

References:

-   [OWASP testing for account enumeration and guessable user account](https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/03-Identity_Management_Testing/04-Testing_for_Account_Enumeration_and_Guessable_User_Account)
-   [OWASP A2:2017-Broken Enumeration](https://owasp.org/www-project-top-ten/2017/A2_2017-Broken_Authentication)

**Comment by Firstyear** [comment]
So @yaleman and I had a longer discussion about this outside of github. We discussed some of the potential attack risks such as:

* Ability to enum accounts allows discovery of identities which may allow for social engineering such as phoning a service desk and requesting a password reset
* Due to how negotiation works we offer MFA first then Password, so for Password only accounts it's immediately obvious they don't have MFA 

We are certain there are more examples, but in each of these situations there are other mitigations or attack vectors.

For example,  discovery of an identity that works in an organisation may occur through a public employee listing or social medial, meaning that social engineering is still possible. This indicates a requirement to ensure that our password reset flows can account for this kind of attack especially in high value accounts. 

Anonymous searching is a default enabled feature, meaning that you don't need to enumerate via authentication, but can just search the directory. Restrictions on what anonymous can read is already implemented, but we can extend this with limits on IPs that can bind as anonymous, https://github.com/kanidm/kanidm/issues/611 or by disabling anonymous authentication outright (may require changes to unixd in it's current form). 

Passwords are already known as a potential weakness for any account which is password only, and extension to account policy to enforce MFA can assist here. For things like service accounts, we can also potentially enforce that api pw's or tokens are always server side generated, not human created to reduce their weakness. Finally, we already have rate limiting in place as per NIST guideliness up to 100 attempts a day, as well as complexity checking through zxcvbn and badlisting of known weak or compromised passwords. This reduces an attackers ability to successfully execute a bruteforce attack, or other types of password attacks that exist today. 

EDIT: After speaking to someone else it's also worth noting that the technical barriers to effectively blocking account enumeration are extremely high. Due to how authentication works in Kani we have to negotiate the methods available. If we were to "pretend" a user account existed on an invalid ID, to simulate a "real" account we'd also need to simulate the rate limit and locking process which would consume memory of the server. Even with/without that, the fact it's in memory and not coming from the entry cache would open the possibility of timing attacks to determine an account name is valid or invalid based on the amount of time to server response. Generally, the further you look into it, the harder and harder it is to actually prevent this. It would be an extraordinary cost to prevent, for what gain? As an "attack" this is not a serious issue when you have other meaningful protections in place. 

When we look at major providers like microsoft, gmail/google, github and others they do NOT try to prevent account/username enumeration. These providers have some of the best security teams in the world, and again, the security of an account is not defined through obscurity of it's name or existence but from other elements. 

More broadly, I think that Kerckhoffs's is important to consider here. As a general view this can be seen as "knowledge of details of the system, should know weaken the system". For example, here on github my account name and potentially even what I have access to (this repo for example), does not weaken my security on github as an attacker is unable to authenticate without the secrets that I posses. With this in mind, I would rather we consider approaches where we assume an attacker *does* know users account names and the roles they may have, and we implemented effective controls to protect users in these situations.
